src/org/labkey/targetedms/parser/speclib/LibSpectrumReader.java

```diff
@@ -0,0 +1,200 @@
+        }
+    }
+
+    private static final Logger LOG = LogManager.getLogger(LibSpectrumReader.class);
+
+    @Nullable
+    public LibSpectrum getLibSpectrum(Container container, Path libPath,
+                                      String modifiedPeptide, int charge) throws SQLException, DataFormatException
+    {
+        String localLibPath = getLocalLibPath(container, libPath);
+        if (null == localLibPath)
+            return null;
+
+        if(!(new File(localLibPath).exists()))
+            return null;
+
+        try (Connection conn = getLibConnection(localLibPath))
+        {
+            return readSpectrum(conn, modifiedPeptide, charge, libPath);
+        }
+    }
+
+    @NotNull
+    public List<LibrarySpectrumMatchGetter.PeptideIdRtInfo> getRetentionTimes(Container container, Path libPath, String modifiedPeptide)
+    {
+        String libFilePath = getLocalLibPath(container, libPath);
+        if (null == libFilePath)
+            return Collections.emptyList();
+
+        // We know it's local file and string may need encoding to convert to Path
+        if(!(new File(libFilePath).exists()))
+        {
+            LOG.debug("File not found: " + libFilePath + ", referenced from container " + container.getPath());
+            return Collections.emptyList();
+        }
+
+        try (Connection conn = getLibConnection(libFilePath))
+        {
+            return readRetentionTimes(conn, modifiedPeptide, libFilePath);
+        }
+        catch(SQLException e)
+        {
+            throw new RuntimeSQLException(e);
+        }
+    }
+
+    protected abstract LibSpectrum readSpectrum(Connection conn, String modifiedPeptide, int charge, Path libPath) throws DataFormatException, SQLException;
+
+    @NotNull
+    protected abstract List<LibrarySpectrumMatchGetter.PeptideIdRtInfo> readRetentionTimes(Connection conn, String modifiedPeptide, String libPath) throws SQLException;
+
+
+    private static final int LIBCACHE_LIMIT = 1000;
+    private static final long LIBCACHE_LIFETIME = CacheManager.DAY;
+
+    private static String getLibCacheKey(@NotNull Container container, String pathStr)
+    {
+        return container.getId() + "::" + pathStr;
+    }
+
+    private static final CacheLoader<String, String> _libCacheLoader = (key, arg) ->
+    {
+        if (null != arg)
+        {
+            Pair<Container, Path> pair = (Pair<Container, Path>) arg;
+            Container container = pair.first;
+            Path remotePath = pair.second;
+            if (null != container && null != remotePath)
+            {
+                File file = LocalDirectory.copyToContainerDirectory(container, remotePath, LOG);
+                if (null != file)
+                    return file.getAbsolutePath();
+            }
+        }
+        return null;
+    };
+
+    private static final BlockingCache<String, String> _libCache =
+            new BlockingCache<>(CacheManager.getStringKeyCache(LIBCACHE_LIMIT, LIBCACHE_LIFETIME, "" +
```

Comment:
Nit - this string concatenation with the empty string can be removed.

Reply:
Done